packages/babel-core/src/transformation/file/options/build-config-chain.js

```diff
@@ -22,27 +23,100 @@ function exists(filename) {
 }
 
 export default function buildConfigChain(opts: Object = {}) {
-  const filename = opts.filename;
-  const builder = new ConfigChainBuilder();
-
-  builder.mergeConfig({
-    options: opts,
-    alias: "base",
-    dirname: process.cwd(),
-  });
-
-  // resolve all .babelrc files
-  if (opts.babelrc !== false) {
-    builder.findConfigs(filename);
+  const filename = opts.filename ? path.resolve(opts.filename) : null;
+  const builder = new ConfigChainBuilder(filename);
+
+  try {
+    builder.mergeConfig({
+      options: opts,
+      alias: "base",
+      dirname: process.cwd(),
+    });
+
+    // resolve all .babelrc files
+    if (opts.babelrc !== false) {
+      builder.findConfigs(filename);
+    }
+  } catch (e) {
+    if (e.code !== "BABEL_IGNORED_FILE") throw e;
+
+    return null;
   }
 
   return builder.configs.reverse();
 }
 
 class ConfigChainBuilder {
-  constructor() {
+  constructor(filename) {
     this.resolvedConfigs = [];
     this.configs = [];
+    this.filename = filename;
+    this.possibleDirs = null;
+  }
+
+  /**
+   * Tests if a filename should be ignored based on "ignore" and "only" options.
+   */
+  shouldIgnore(
+    ignore: Array<string | RegExp | Function>,
+    only?: Array<string | RegExp | Function>,
+    dirname: string,
+  ): boolean {
+    if (!this.filename) return false;
+
+    if (ignore) {
+      if (!Array.isArray(ignore)) throw new Error(".ignore should be an array.");
```

Comment:
Good question. I think I'll likely be refactoring this code a ton anyway moving forward, happy if we address this then?